File: tools/budgets/internal/budgets/check.go
// check —— 守恒校验与延迟负债表（spec §3.6）。
//
// 守恒律：ΣP95 − 并行重叠 ≤ total_p95_budget（基准 configs/budgets/latency.yaml，
// 当前 1500ms）；违反 → exit 20，并输出「延迟负债表」：各段实际值 vs 预算值、
// 差值、超标段（只认 P95，p50 不参与守恒计算）。
package budgets

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"

	"gopkg.in/yaml.v3"
)

// InputError 表示预算配置 / 延迟报告 / history 不可读或不符合 schema（CLI exit 2）。
type InputError struct{ msg string }

func (e *InputError) Error() string { return e.msg }

func inputErrorf(format string, args ...any) error {
	return &InputError{msg: fmt.Sprintf(format, args...)}
}

// SegmentBudget 是 configs/budgets/latency.yaml 中一段的预算（只认 P95）。
type SegmentBudget struct {
	ID    string  `yaml:"id"`
	Asset string  `yaml:"asset"`
	P50   float64 `yaml:"p50"`
	P95   float64 `yaml:"p95"`
	Note  string  `yaml:"note"`
}

// BudgetConfig 是延迟预算基准：总 P95 预算 + 各段预算（rules 等说明键不参与计算）。
type BudgetConfig struct {
	TotalP95Budget float64         `yaml:"total_p95_budget"`
	Segments       []SegmentBudget `yaml:"segments"`
}

// LoadConfig 经 yaml.v3 解析 latency.yaml 并校验非负性与完整性。
func LoadConfig(path string) (BudgetConfig, error) {
	var config BudgetConfig
	data, err := os.ReadFile(path)
	if err != nil {
		return config, inputErrorf("预算配置不存在或不可读: %s", path)
	}
	if err := yaml.Unmarshal(data, &config); err != nil {
		return config, inputErrorf("预算配置不可解析: %v", err)
	}
	if math.IsNaN(config.TotalP95Budget) || config.TotalP95Budget < 0 {
		return config, inputErrorf("total_p95_budget 须为非负数: %v", config.TotalP95Budget)
	}
	if len(config.Segments) == 0 {
		return config, inputErrorf("segments 为空")
	}
	for _, seg := range config.Segments {
		if seg.ID == "" {
			return config, inputErrorf("segment 缺 id: %+v", seg)
		}
		if math.IsNaN(seg.P95) || seg.P95 < 0 {
			return config, inputErrorf("segment %s 的 p95 须为非负数: %v", seg.ID, seg.P95)
		}
	}
	return config, nil
}

// SegmentSample 是报告中一段的实测值。
type SegmentSample struct {
	ID  string  `json:"id"`
	P50 float64 `json:"p50"`
	P95 float64 `json:"p95"`
}

// LatencyReport 是单份夜间延迟报告（格式见包 doc 注释）。
type LatencyReport struct {
	Commit    string          `json:"commit"`
	Timestamp string          `json:"timestamp"`
	OverlapMS float64         `json:"overlap_ms"` // 缺省 0
	Segments  []SegmentSample `json:"segments"`
}

func loadReport(path string) (LatencyReport, error) {
	var report LatencyReport
	data, err := os.ReadFile(path)
	if err != nil {
		return report, inputErrorf("延迟报告不存在或不可读: %s", path)
	}
	if err := json.Unmarshal(data, &report); err != nil {
		return report, inputErrorf("延迟报告不可解析: %v", err)
	}
	return report, nil
}

// SegmentRow 是负债表一行：段预算 vs 实际。
type SegmentRow struct {
	ID        string
	BudgetP95 float64
	ActualP95 float64
	Delta     float64 // 实际 − 预算
	Over      bool    // 实际 P95 > 该段预算
}

// CheckResult 是守恒校验结果与负债表数据。
type CheckResult struct {
	Commit      string
	TotalActual float64 // ΣP95
	OverlapMS   float64
	EffectiveMS float64 // ΣP95 − 并行重叠
	BudgetMS    float64
	OK          bool
	OverByMS    float64
	Rows        []SegmentRow
}

// OverSegments 返回实际 P95 超过自身段预算的段 id（负债表「超标段」清单）。
func (r CheckResult) OverSegments() []string {
	var over []string
	for _, row := range r.Rows {
		if row.Over {
			over = append(over, row.ID)
		}
	}
	return over
}

// Evaluate 对照预算基准评估单份报告：段集合须与配置完全一致，p95/overlap 非负。
func Evaluate(report LatencyReport, config BudgetConfig) (CheckResult, error) {
	var result CheckResult
	if len(report.Segments) == 0 {
		return result, inputErrorf("报告缺 segments 列表")
	}
	actual := make(map[string]float64, len(report.Segments))
	for _, seg := range report.Segments {
		if seg.ID == "" {
			return result, inputErrorf("segment 缺 id: %+v", seg)
		}
		if _, dup := actual[seg.ID]; dup {
			return result, inputErrorf("segment id 重复: %s", seg.ID)
		}
		if math.IsNaN(seg.P95) || seg.P95 < 0 {
			return result, inputErrorf("segment %s 的 p95 须为非负数: %v", seg.ID, seg.P95)
		}
		actual[seg.ID] = seg.P95
	}
	budgetByID := make(map[string]float64, len(config.Segments))
	var missing, unknown []string
	for _, seg := range config.Segments {
		budgetByID[seg.ID] = seg.P95
		if _, ok := actual[seg.ID]; !ok {
			missing = append(missing, seg.ID)
		}
	}
	for _, seg := range report.Segments {
		if _, ok := budgetByID[seg.ID]; !ok {
			unknown = append(unknown, seg.ID)
		}
	}
	if len(missing) > 0 || len(unknown) > 0 {
		return result, inputErrorf("报告段与预算配置不一致: 缺 %v，多 %v", missing, unknown)
	}
	if math.IsNaN(report.OverlapMS) || report.OverlapMS < 0 {
		return result, inputErrorf("overlap_ms 须为非负数: %v", report.OverlapMS)
	}

	result = CheckResult{Commit: report.Commit, OverlapMS: report.OverlapMS, BudgetMS: config.TotalP95Budget}
	if result.Commit == "" {
		result.Commit = "?"
	}
	for _, seg := range config.Segments {
		p95 := actual[seg.ID]
		delta := p95 - seg.P95
		result.TotalActual += p95
		result.Rows = append(result.Rows, SegmentRow{ID: seg.ID, BudgetP95: seg.P95, ActualP95: p95, Delta: delta, Over: delta > 0})
	}
	result.EffectiveMS = result.TotalActual - result.OverlapMS
	result.OK = result.EffectiveMS <= result.BudgetMS
	result.OverByMS = math.Max(0, result.EffectiveMS-result.BudgetMS)
	return result, nil
}

// fmtMS 毫秒展示：整数去小数点，非整数保留 1 位。
func fmtMS(v float64) string {
	if v == math.Trunc(v) && !math.IsInf(v, 0) && !math.IsNaN(v) {
		return strconv.FormatFloat(v, 'f', 0, 64)
	}
	return strconv.FormatFloat(v, 'f', 1, 64)
}

// FormatDebtTable 渲染延迟负债表：段/预算/实际/差值/状态 + 超标段清单 + 守恒判定。
func FormatDebtTable(result CheckResult, reportPath string) string {
	where := ""
	if reportPath != "" {
		where = "，报告=" + reportPath
	}
	lines := []string{
		fmt.Sprintf("延迟负债表（commit=%s%s）", result.Commit, where),
		fmt.Sprintf("%-14s%10s%10s%8s  状态", "段", "预算P95", "实际P95", "差值"),
	}
	for _, row := range result.Rows {
		delta := fmtMS(row.Delta)
		if row.Delta > 0 {
			delta = "+" + delta
		}
		status := "正常"
		if row.Over {
			status = "超标"
		}
		lines = append(lines, fmt.Sprintf("%-14s%10s%10s%8s  %s",
			row.ID, fmtMS(row.BudgetP95), fmtMS(row.ActualP95), delta, status))
	}
	lines = append(lines, strings.Repeat("─", 58))
	lines = append(lines, fmt.Sprintf("ΣP95=%s 并行重叠=%s 有效总延迟=%s 总预算=%s 超支=%s",
		fmtMS(result.TotalActual), fmtMS(result.OverlapMS), fmtMS(result.EffectiveMS), fmtMS(result.BudgetMS), fmtMS(result.OverByMS)))
	if over := result.OverSegments(); len(over) > 0 {
		lines = append(lines, "超标段: "+strings.Join(over, ", "))
	}
	if result.OK {
		lines = append(lines, "守恒校验通过：ΣP95−并行重叠 ≤ 总预算")
	} else {
		lines = append(lines, fmt.Sprintf("守恒校验违反：ΣP95−并行重叠=%sms > 总预算%sms，超支 %sms",
			fmtMS(result.EffectiveMS), fmtMS(result.BudgetMS), fmtMS(result.OverByMS)))
	}
	return strings.Join(lines, "\n")
}

File: tools/budgets/src/budgets/check.py
"""budgets.check —— 守恒校验与延迟负债表（spec §3.6）。

守恒律：ΣP95 − 并行重叠 ≤ total_p95_budget（基准 configs/budgets/latency.yaml，
当前 1500ms）；违反 → exit 20，并输出「延迟负债表」：各段实际值 vs 预算值、
差值、超标段（只认 P95，p50 不参与守恒计算）。

latency.json 输入格式（单份夜间报告，JSON 对象）：
    {
      "commit": "a1b2c3",                     # 产生该报告的 commit（展示用）
      "timestamp": "2026-08-28T00:00:00Z",   # ISO-8601（展示用）
      "overlap_ms": 50,                       # 可选：并行段重叠（默认 0，非负）
      "segments": [                           # 段 id 须与预算配置一致
        {"id": "tail_silence", "p50": 450, "p95": 600},
        ...
      ]
    }

latency.yaml 是受控扁平子集（顶层标量 + 内联 mapping 列表，值不含逗号/花括号），
用行级极简解析器读取，避免为固定 schema 引入 PyYAML 依赖。"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path


class BudgetInputError(ValueError):
    """预算配置或延迟报告不可读/不符合 schema。"""


def _is_number(value) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def fmt_ms(value: float | None) -> str:
    """毫秒展示：整数去小数点，非整数保留 1 位，缺失用 —。"""
    if value is None:
        return "—"
    return str(int(value)) if float(value).is_integer() else f"{value:.1f}"


def _scalar(token: str):
    token = token.strip().strip("'\"")
    if token in ("", "~", "null") or token.lower() in ("nan", "inf", "-inf", "infinity"):
        return None
    for cast in (int, float):
        try:
            return cast(token)
        except ValueError:
            pass
    return {"true": True, "false": False}.get(token, token)


@dataclass(frozen=True)
class SegmentBudget:
    id: str
    p95: float


@dataclass(frozen=True)
class BudgetConfig:
    total_p95_budget: float
    segments: tuple[SegmentBudget, ...]


def load_config(path: Path) -> BudgetConfig:
    """解析 latency.yaml（total_p95_budget + segments 内联 mapping 列表）。"""
    if not path.is_file():
        raise BudgetInputError(f"预算配置不存在: {path}")
    try:
        lines = path.read_text(encoding="utf-8").splitlines()
    except OSError as exc:
        raise BudgetInputError(f"预算配置不可读: {exc}") from exc

    total, segments, in_segments = None, [], False
    for raw in lines:
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        if raw[:1] not in (" ", "\t"):  # 顶层键
            key, sep, value = raw.partition(":")
            in_segments = bool(sep) and key.strip() == "segments"
            if sep and key.strip() == "total_p95_budget":
                total = _scalar(value)
            continue
        if not (in_segments and stripped.startswith("- ")):
            continue  # rules 等其它列表项不参与计算
        item = stripped[2:].strip()
        if not (item.startswith("{") and item.endswith("}")):
            continue
        mapping = {}
        for part in item[1:-1].split(","):
            key, sep, value = part.partition(":")
            if sep and key.strip():
                mapping[key.strip()] = _scalar(value)
        seg_id, p95 = mapping.get("id"), mapping.get("p95")
        if not isinstance(seg_id, str) or not seg_id:
            raise BudgetInputError(f"segment 缺 id: {mapping}")
        if not _is_number(p95) or p95 < 0:
            raise BudgetInputError(f"segment {seg_id} 的 p95 须为非负数: {p95!r}")
        segments.append(SegmentBudget(seg_id, p95))

    if not _is_number(total) or total < 0:
        raise BudgetInputError(f"total_p95_budget 须为非负数: {total!r}")
    if not segments:
        raise BudgetInputError("segments 为空")
    return BudgetConfig(total, tuple(segments))


def load_report(path: Path) -> dict:
    if not path.is_file():
        raise BudgetInputError(f"延迟报告不存在: {path}")
    try:
        report = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise BudgetInputError(f"延迟报告不可读: {exc}") from exc
    if not isinstance(report, dict):
        raise BudgetInputError("延迟报告必须是 JSON 对象")
    return report


@dataclass
class SegmentRow:
    id: str
    budget_p95: float
    actual_p95: float
    delta: float
    status: str  # 正常 | 超标


@dataclass
class CheckResult:
    commit: str
    total_actual: float  # ΣP95
    overlap_ms: float
    effective_ms: float  # ΣP95 − 并行重叠
    budget_ms: float
    ok: bool
    over_by_ms: float
    rows: list[SegmentRow]

    @property
    def over_segments(self) -> list[str]:
        return [row.id for row in self.rows if row.status == "超标"]


def evaluate(report: dict, config: BudgetConfig) -> CheckResult:
    segments = report.get("segments")
    if not isinstance(segments, list) or not segments:
        raise BudgetInputError("报告缺 segments 列表")
    actual: dict[str, float] = {}
    for seg in segments:
        seg_id = seg.get("id") if isinstance(seg, dict) else None
        if not isinstance(seg_id, str) or not seg_id:
            raise BudgetInputError(f"segment 缺 id: {seg!r}")
        p95 = seg.get("p95") if isinstance(seg, dict) else None
        if seg_id in actual:
            raise BudgetInputError(f"segment id 重复: {seg_id}")
        if not _is_number(p95) or p95 < 0:
            raise BudgetInputError(f"segment {seg_id} 的 p95 须为非负数: {p95!r}")
        actual[seg_id] = float(p95)

    budget_by_id = {seg.id: seg.p95 for seg in config.segments}
    missing = [seg_id for seg_id in budget_by_id if seg_id not in actual]
    unknown = [seg_id for seg_id in actual if seg_id not in budget_by_id]
    if missing or unknown:
        raise BudgetInputError(f"报告段与预算配置不一致: 缺 {missing or '无'}，多 {unknown or '无'}")
    overlap = report.get("overlap_ms", 0)
    if not _is_number(overlap) or overlap < 0:
        raise BudgetInputError(f"overlap_ms 须为非负数: {overlap!r}")

    rows = []
    for seg in config.segments:
        delta = actual[seg.id] - seg.p95
        rows.append(SegmentRow(seg.id, seg.p95, actual[seg.id], delta,
                               "超标" if delta > 0 else "正常"))

    total_actual = sum(actual.values())
    effective = total_actual - float(overlap)
    ok = effective <= config.total_p95_budget
    return CheckResult(
        commit=str(report.get("commit", "?")),
        total_actual=total_actual,
        overlap_ms=float(overlap),
        effective_ms=effective,
        budget_ms=config.total_p95_budget,
        ok=ok,
        over_by_ms=max(0.0, effective - config.total_p95_budget),
        rows=rows,
    )


def format_debt_table(result: CheckResult, report_path: str = "") -> str:
    where = f"，报告={report_path}" if report_path else ""
    lines = [f"延迟负债表（commit={result.commit}{where}）",
             f"{'段':<14}{'预算P95':>10}{'实际P95':>10}{'差值':>8}  状态"]
    for row in result.rows:
        delta = ("+" if row.delta > 0 else "") + fmt_ms(row.delta)
        lines.append(f"{row.id:<14}{fmt_ms(row.budget_p95):>10}{fmt_ms(row.actual_p95):>10}"
                     f"{delta:>8}  {row.status}")
    lines.append("─" * 58)
    lines.append(f"ΣP95={fmt_ms(result.total_actual)} 并行重叠={fmt_ms(result.overlap_ms)}"
                 f" 有效总延迟={fmt_ms(result.effective_ms)} 总预算={fmt_ms(result.budget_ms)}"
                 f" 超支={fmt_ms(result.over_by_ms)}")
    if result.over_segments:
        lines.append(f"超标段: {', '.join(result.over_segments)}")
        lines.append(f"守恒校验违反：ΣP95−并行重叠={fmt_ms(result.effective_ms)}ms > 总预算"
                     f"{fmt_ms(result.budget_ms)}ms，超支 {fmt_ms(result.over_by_ms)}ms")
    else:
        lines.append("守恒校验通过：ΣP95−并行重叠 ≤ 总预算")
    return "\n".join(lines)
